In [1]:
# Importing all the Required Libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
from plotly.offline import init_notebook_mode
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import seaborn as sns
import datetime as dt
import warnings
import plotly.graph_objects as go
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns',None)
init_notebook_mode(connected=True)

In [2]:
df=pd.read_csv(r"C:\Users\91940\Downloads\71_Centuries_of_Virat_Kohli.csv")

In [3]:
df.head()


Score Out/Not Out       Against  Batting Order  Inn.  Strike Rate  \
0    116         Out     Australia              6     2          NaN   
1    103         Out   New Zealand              5     2          NaN   
2    103         Out       England              5     2          NaN   
3    107         Out     Australia              5     2          NaN   
4    119         Out  South Africa              4     1          NaN   

                                  Venue        Column1   H/A        Date  \
0                         Adelaide Oval       Adelaide  Away  24-01-2012   
1                M. Chinnaswamy Stadium      Bangalore  Home  31-08-2012   
2  Vidarbha Cricket Association Stadium         Nagpur  Home  13-12-2012   
3             M. A. Chidambaram Stadium        Chennai  Home  22-02-2013   
4                     Wanderers Stadium   Johannesburg  Away  18-12-2013   

  Result Format Man of the Match Captain  Unnamed: 14  
0   Lost   Test               No      No          NaN  
1    Won   Test              Yes      No          NaN  
2  Drawn   Test               No      No          NaN  
3    Won   Test               No      No          NaN  
4  Drawn   Test               No      No          NaN

In [4]:
df.drop('Unnamed: 14',axis=1,inplace=True)


In [5]:
df.head()


Score Out/Not Out       Against  Batting Order  Inn.  Strike Rate  \
0    116         Out     Australia              6     2          NaN   
1    103         Out   New Zealand              5     2          NaN   
2    103         Out       England              5     2          NaN   
3    107         Out     Australia              5     2          NaN   
4    119         Out  South Africa              4     1          NaN   

                                  Venue        Column1   H/A        Date  \
0                         Adelaide Oval       Adelaide  Away  24-01-2012   
1                M. Chinnaswamy Stadium      Bangalore  Home  31-08-2012   
2  Vidarbha Cricket Association Stadium         Nagpur  Home  13-12-2012   
3             M. A. Chidambaram Stadium        Chennai  Home  22-02-2013   
4                     Wanderers Stadium   Johannesburg  Away  18-12-2013   

  Result Format Man of the Match Captain  
0   Lost   Test               No      No  
1    Won   Test              Yes      No  
2  Drawn   Test               No      No  
3    Won   Test               No      No  
4  Drawn   Test               No      No

In [6]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Score             71 non-null     int64  
 1   Out/Not Out       71 non-null     object 
 2   Against           71 non-null     object 
 3   Batting Order     71 non-null     int64  
 4   Inn.              71 non-null     int64  
 5   Strike Rate       44 non-null     float64
 6   Venue             71 non-null     object 
 7   Column1           71 non-null     object 
 8   H/A               71 non-null     object 
 9   Date              71 non-null     object 
 10  Result            71 non-null     object 
 11  Format            71 non-null     object 
 12  Man of the Match  71 non-null     object 
 13  Captain           71 non-null     object 
dtypes: float64(1), int64(3), object(10)
memory usage: 7.9+ KB


In [7]:
#EDA

In [8]:
df.head()


Score Out/Not Out       Against  Batting Order  Inn.  Strike Rate  \
0    116         Out     Australia              6     2          NaN   
1    103         Out   New Zealand              5     2          NaN   
2    103         Out       England              5     2          NaN   
3    107         Out     Australia              5     2          NaN   
4    119         Out  South Africa              4     1          NaN   

                                  Venue        Column1   H/A        Date  \
0                         Adelaide Oval       Adelaide  Away  24-01-2012   
1                M. Chinnaswamy Stadium      Bangalore  Home  31-08-2012   
2  Vidarbha Cricket Association Stadium         Nagpur  Home  13-12-2012   
3             M. A. Chidambaram Stadium        Chennai  Home  22-02-2013   
4                     Wanderers Stadium   Johannesburg  Away  18-12-2013   

  Result Format Man of the Match Captain  
0   Lost   Test               No      No  
1    Won   Test              Yes      No  
2  Drawn   Test               No      No  
3    Won   Test               No      No  
4  Drawn   Test               No      No

In [9]:
df['year']=pd.to_datetime(df['Date'])
df['year']=df['year'].dt.year
df['Date']=df['Date'].replace('08-09-2022','09-08-2022')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Score             71 non-null     int64  
 1   Out/Not Out       71 non-null     object 
 2   Against           71 non-null     object 
 3   Batting Order     71 non-null     int64  
 4   Inn.              71 non-null     int64  
 5   Strike Rate       44 non-null     float64
 6   Venue             71 non-null     object 
 7   Column1           71 non-null     object 
 8   H/A               71 non-null     object 
 9   Date              71 non-null     object 
 10  Result            71 non-null     object 
 11  Format            71 non-null     object 
 12  Man of the Match  71 non-null     object 
 13  Captain           71 non-null     object 
 14  year              71 non-null     int64  
dtypes: float64(1), int64(4), object(10)
memory usage: 8.4+ KB


In [11]:
df.describe()


Score  Batting Order       Inn.  Strike Rate         year
count   71.000000      71.000000  71.000000    44.000000    71.000000
mean   132.140845       3.521127   1.732394   114.019545  2015.225352
std     35.911119       0.714326   0.675230    25.257567     2.894316
min    100.000000       1.000000   1.000000    84.900000  2009.000000
25%    107.000000       3.000000   1.000000    96.632500  2013.000000
50%    119.000000       3.000000   2.000000   108.935000  2016.000000
75%    139.500000       4.000000   2.000000   120.787500  2018.000000
max    254.000000       6.000000   4.000000   200.000000  2022.000000

In [13]:
df.groupby(['Against','year'],as_index=False)['Score'].max().sort_values(by='Score',ascending=False).reset_index(drop=True).head()


Against  year  Score
0  South Africa  2019    254
1     Sri Lanka  2017    243
2       England  2016    235
3   New Zealand  2016    211
4    Bangladesh  2017    204

### Observation:
    Kohli Scores Highest (254) Runs Against South Africa in the year 2019 and against Sri Lanka in 2017


In [14]:
fig=px.line(df,x='Date',y='Score',markers=True,template='simple_white',title='<b>Kohli\'s hundreds over time</b>')
fig.update_layout(title_x=0.5)

In [15]:
fig=px.bar(df.groupby('Against',as_index=False)['Score'].count().sort_values(by='Score',ascending=False).reset_index(drop=True),x='Against',y='Score',color='Against',text='Score',labels={'Score':'Count'},template='simple_white',title='<b>Kohli\'s Centuries Against Opponents</b>')
fig.update_layout(title_x=0.5)
fig.show()
fig=px.bar(df.groupby(['Against','H/A'],as_index=False)['Score'].count().sort_values(by='Score',ascending=False).reset_index(drop=True),x='Against',y='Score',color='H/A',text='Score',labels={'Score':'Count'},template='simple_white',title='<b>Kohli\'s Centuries Analysis by Venue Home/Away</b>')
fig.update_layout(title_x=0.5,legend=dict(orientation='h',yanchor='bottom',y=1.02,xanchor='right',x=1))

### Observation:
From the Above graph, Virat Kohli Scores more Centuries against Australia(15) , Srilanka(13) and West Indies(11).

He Score (9) centuries from Away ground against Austalia and (8) Centuries against Srilanka.

In Home Ground he Scores (6) centuries against Australia, west indies and New Zealand.

In [16]:
fig=px.bar(df.groupby('year',as_index=False)['Score'].count(),x='year',y='Score',text='Score',labels={'Score':'Count'},color_discrete_sequence=['lightgreen'],template='seaborn',title='<b>Virat Kohli\'s hundreds year-by-year</b>')
fig.update_layout(title_x=0.5)
fig.show()

### Observation:

In the Year 2016 and 2017 Virat Kohli Scored 11 centuries and he is the secon person after Sachin Tendulkar

In 1998 Sachin Tendulkar scored 12 centuries in all formats combined, which is the record for Most centuries in the calender year in cricket in all formats.

In [17]:
fig=px.bar(df.groupby(['Against','Out/Not Out'],as_index=False)['Score'].count(),x='Against',y='Score',color='Out/Not Out',color_discrete_sequence=['olive',' #F8766D'],text='Score',labels={'Out/Not Out':'result','Score':'Count'},template='ggplot2',title='<b>Whether Virat Out or Not Out when he Hit a Century</b>')
fig.update_layout(title_x=0.5,legend=dict(orientation='h',yanchor='bottom',y=1.02,xanchor='right',x=1))
fig.show()
fig=px.pie(df.groupby('Out/Not Out',as_index=False)['Score'].count().sort_values(by='Score',ascending=False).reset_index(drop=True),names='Out/Not Out',values='Score',color='Out/Not Out',color_discrete_sequence=[' #F8766D','olive'],labels={'Out/Not Out':'result','Score':'Count'},template='seaborn',hole=0.5,title='<b>Out or Not Out')
fig.update_layout(title_x=0.5,legend=dict(orientation='h',yanchor='bottom',y=1.02,xanchor='right',x=1))
fig.show()

### Observation:
    Out of 71 Centuries, Kohli outs 51 times

    He outs mostly for Australia i.e 13 times

In [18]:
fig=px.sunburst(df.groupby(['Format','Against','Result'],as_index=False)['Score'].count(),path=['Format','Result','Against'],values='Score',color='Result',title='<b>Match results when kohli scores centuries against opponents')
fig.update_layout(title_x=0.5)

In [19]:
# Total Centuries scored in different Formats
df.groupby(['Format'],as_index=False)['Score'].count().sort_values(by='Score',ascending = False).reset_index(drop=True)

Format  Score
0    ODI     43
1   Test     27
2   T20I      1

In [20]:
fig=px.bar(df.groupby(['Format','Result'],as_index=False)['Score'].count(),x='Format',y='Score',color='Result',text='Score',labels={'Out/Not Out':'result','Score':'Count'},template='seaborn',title='<b>Match Result Based on Virat\'s Centuries</b>')
fig.update_layout(title_x=0.5,legend=dict(orientation='h',yanchor='bottom',y=1.02,xanchor='right',x=1))
fig.show()

### Observation:
    From the above Analysis, Centuries scored by Virat in different formats are (43) in ODI, (27) in Test and (1) in T20I

    when Kohli scored centuries, india won (34) ODI matches, (13) Test Matches and (1) T20I Matches

In [21]:
df.groupby(['Captain'],as_index=False)['Score'].count()


Captain  Score
0      No     30
1     Yes     41

In [22]:
fig=px.bar(df.groupby(['Captain','Result'],as_index=False)['Score'].count(),x='Captain',y='Score',color='Result',text='Score',labels={'Out/Not Out':'result','Score':'Count'},template='plotly',title='<b>Match Result Based on Virat\'s Century when he was Captain</b>')
fig.update_layout(title_x=0.5,legend=dict(orientation='h',yanchor='bottom',y=1.02,xanchor='right',x=1))
fig.show()

### Observation:
From the Above graph, when Virat was a Captain he Scores (40) Centuries and (30) centuries without captaincy

In Kohli's captaincy, when he scored centuries india won 27 matches and lost 9 matches

In [23]:
fig = make_subplots(rows=1, cols=2, specs=[[{"type": "pie"}, {"type": "pie"}]],subplot_titles=('<b>Man of the Match', '<b>Captain'))
colors = ['crimson','blue']
fig.add_trace(go.Pie(values = df.groupby('Man of the Match',as_index=False)['Score'].count().Score, labels = df.groupby('Man of the Match',as_index=False)['Score'].count()['Man of the Match'], hole = 0.4,marker_colors = colors,name='Man of match'),row=1,col=1)
fig.add_trace(go.Pie(values = df.groupby('Captain',as_index=False)['Score'].count().Score, labels = df.groupby('Captain',as_index=False)['Score'].count()['Captain'], hole = 0.4,marker_colors = colors,name='Captain'),row=1,col=2)
fig.update_layout(showlegend=True)

### Observation:
Out of 71 Centuries Scored by Virat, he became Man of the Match 38 times

As a captian, Virat Scored 41 centuries

In [24]:
fig=px.pie(df.groupby(['Batting Order'],as_index=False)['Score'].count(),values = 'Score', names = 'Batting Order', hole = 0.4,template='seaborn',labels={'Score':'Count'},title='<b>Virat\'s Centuries based on Batting Order')
fig.update_layout(title_x=0.5,legend=dict(orientation='h',yanchor='bottom',y=1.02,xanchor='right',x=1))

### Observation:
    Virat Kohli's best and most used Batting order is 3 and 4



In [26]:
fig=px.bar(df.groupby(['Against','Batting Order'],as_index=False)['Score'].count().sort_values(by='Score',ascending=False).reset_index(drop=True),x='Against',y='Score',color='Batting Order',text='Score',labels={'Score':'Count'},template='simple_white',title='<b>Kohli\'s Centuries Against opponents based on batting order</b>')
fig.update_layout(title_x=0.5)

In [27]:
fig=px.pie(df.groupby(['H/A'],as_index=False)['Score'].count(),values = 'Score', names = 'H/A', hole = 0.4,color_discrete_sequence=['Salmon','Turquoise'],title='<b>Total Centuries of kohli in Away & Home ground')
fig.update_layout(title_x=0.5,legend=dict(orientation='h',yanchor='bottom',y=1.02,xanchor='right',x=1))

### Observation:

Virat Kohli Scored 39 centuries in Home ground and 32 centuries in away ground



In [28]:
fig=px.treemap(df,path=[px.Constant('Virat Kohli\'s centuries'),'Against','Venue','Score'],hover_name='Against',hover_data=['year'],template='ggplot2',title='<b>TreeMap of Kohli\'s centuries in different stadiums')
fig.update_traces(root_color='lightgrey')

In [29]:
fig=px.choropleth(df.groupby('Against',as_index=False)['Score'].count(),locations='Against',locationmode='country names',color='Score',labels={'Score':'Count'},hover_data=['Score'],template='plotly',title='<b>Map View of Kohli\'s Centuries against opponents')
fig.update_layout(title_x=0.5)

In [30]:
a=df
a['Date']=pd.to_datetime(a['Date'])
a.sort_values(by='Date',ascending=True,inplace=True)
a['Century']=[i for i in range(1,72)]
fig=px.line(a,x='Date',y='Century',markers=True,template='plotly_dark',color_discrete_sequence=['magenta'],title='<b>Total Centuries of kohli over time ')
fig.update_layout(title_x=0.5)

### Observation:
    On Thursday (September 8), Virat Kohli scored his long waited 71st international century, which was also his maiden ton in T20I cricket